packages/tools/src/tools/AdvancedMagnifyViewport.ts

```diff
@@ -0,0 +1,614 @@
+import { vec2, vec3 } from 'gl-matrix';
+import {
+  getEnabledElement,
+  eventTarget,
+  utilities as csUtils,
+} from '@cornerstonejs/core';
+import type { Types } from '@cornerstonejs/core';
+import {
+  SegmentationRepresentations,
+  ToolModes,
+  Events as cstEvents,
+} from '../enums';
+import { ToolGroupManager } from '../store';
+import { debounce } from '../utilities';
+import { ToolModeChangedEventType } from '../types/EventTypes';
+import { segmentation } from '..';
+import { EventTypes, IToolGroup } from '../types';
+import { state } from '../store';
+import {
+  AnnotationTool,
+  AdvancedMagnifyTool,
+  SegmentationDisplayTool,
+} from './';
+import { distanceToPoint } from '../utilities/math/point';
+
+const MAGNIFY_CLASSNAME = 'advancedMagnifyTool';
+const MAGNIFY_VIEWPORT_INITIAL_RADIUS = 125;
+
+const isSegmentation = (actor) => actor.uid !== actor.referenceId;
+
+export type AutoPanCallbackData = {
+  points: {
+    currentPosition: {
+      canvas: Types.Point2;
+      world: Types.Point3;
+    };
+    newPosition: {
+      canvas: Types.Point2;
+      world: Types.Point3;
+    };
+  };
+  delta: {
+    canvas: Types.Point2;
+    world: Types.Point3;
+  };
+};
+
+export type AutoPanCallback = (data: AutoPanCallbackData) => void;
+
+class AdvancedMagnifyViewport {
+  private _viewportId: string;
+  private _sourceEnabledElement: Types.IEnabledElement;
+  private _enabledElement: Types.IEnabledElement = null;
+  private _sourceToolGroup: IToolGroup = null;
+  private _magnifyToolGroup: IToolGroup = null;
+  private _isViewportReady = false;
+  private _radius = 0;
+  private _resized = false;
+  private _resizeViewportAsync: () => void;
+  private _canAutoPan = false;
+  private _autoPan: {
+    enabled: boolean;
+    padding: number;
+    callback: AutoPanCallback;
+  };
+  public position: Types.Point2;
+  public zoomFactor: number;
+  public visible: boolean;
+
+  constructor({
+    magnifyViewportId,
+    sourceEnabledElement,
+    radius = MAGNIFY_VIEWPORT_INITIAL_RADIUS,
+    position = [0, 0],
+    zoomFactor,
+    autoPan,
+  }: {
+    magnifyViewportId?: string;
+    sourceEnabledElement: Types.IEnabledElement;
+    radius?: number;
+    position?: Types.Point2;
+    zoomFactor: number;
+    autoPan: {
+      enabled: boolean;
+      padding: number;
+      callback: AutoPanCallback;
+    };
+  }) {
+    // Private properties
+    this._viewportId = magnifyViewportId ?? csUtils.uuidv4();
+    this._sourceEnabledElement = sourceEnabledElement;
+    this._autoPan = autoPan;
+
+    // Pulic properties
+    this.radius = radius;
+    this.position = position;
+    this.zoomFactor = zoomFactor;
+    this.visible = true;
+
+    this._nativeMouseDownCallback = this._nativeMouseDownCallback.bind(this);
```

Comment:
what does native mean in this context?

Reply:
mousedown, mouseup, mousemove, dblclick, etc. (non-cornerstone events)

Reply:
I think browserEvents makes more sense